#273: Cached Injector - wip ready for trashing.

snorrees opened the pull request:
WiP injector cache - names of classes are temporary to ease the diff with existing classes.

Injector is configurable via WorldConfiguration.

* **Injector** - interface (now called i Inject)
* **DefaultInjector** (now called Injector) - more or less the current Injector-class
* **CachedInjector** - caches class, annotation & field annotations in static maps

Todo: Actulal performance benchmark.

This spesific pull-request **should not be merged**, but should serve as a point of discussion.

snorrees commented on artemis/src/main/java/com/artemis/Inject.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ *
+ */
+public interface Inject {
```
Quick and dirty interface - initialize & update can probably be combined.

DaanVanYperen commented:
I'd separate concerns by moving the actual reflection caching into `com.artemis.utils.reflect.ClassReflection`. That way there's a broader caching benefit as well.

Can you say something about the relative performance and memory usage? (take your game as example).

Also consider if the caching mechanic will require any special tests beyond the ones for the vanilla one.

snorrees commented:
The memory usage should be quite minimal, as classes-keys are already in memory, and the cache-values are quite small. I guess if there are hundreds of classes with several fields things might add up, but  it should be negilable relative to the mappers and systems themselves.

If I move stuff to ClassReflection, how should static caches be controlled? Static methods for enabling and/or clearing the caches? 

Furthermore, the way its currently implemented it is very @Wire-centric eg:
```java
    private enum ClassType {
        MAPPER, SYSTEM, MANAGER, FACTORY, CUSTOM
    }

    private enum WireType {
        WIRE, IGNORED
    }

    private static class CachedClass {
        public CachedClass(Class<?> clazz) {
            this.clazz = clazz;
        }
        Class<?> clazz;
        Field[] allFields;
        Wire wireAnnotation;
        boolean injectInherited;
        boolean failOnNull;
        WireType wireType;
    }

    private static class CachedField{
        public CachedField(boolean wire, String name) {
            this.wire = wire;
            this.name = name;
        }

        boolean wire;
        String name;
    }
```

Does it really fit int ClassReflection?

As for performance, I will have to create a jmh test to really test the benefits. In my case, where I inject alot of instances of the same classes (thousands of instances, 25-50 classes), on Android, this gives about a 300-400% speed boost.

DaanVanYperen commented:
The principles you use here can probably be generalized to the ClassReflection. It depends a lot if there's a significant performance to be gained, don't know enough about android. 

Interested to know what time it takes to DI a more typical game on android. If there isn't really a big impact on typical games might make sense to only PR the extensible API and put cached into contrib or in the game.

If memory footprint is tiny and the performance improvement is significant for most games it would make sense to make it a default functionality instead of having two different implementations of injector.

@junkdog with all your io work, could a compile time cache be worth the effort for android? Could be desirable for the networking end of things as well.

Late so pardon the dazed post. ;)

snorrees commented:
Ah - reflection on the annotations. Inception!

DaanVanYperen commented:
@snorrees how's it going!

junkdog commented:
> @junkdog with all your io work, could a compile time cache be worth the effort for android?

I reckon, yes.

@snorrees sorry, been busy - ping me on gitter or here if you haven't heard anything from me before bedtime.

snorrees commented:
Ah sorry, I was unable to work on this yesterday, got sidetracked by a nasty, unrelated bug - I can hopefully get a jhm test for the implementation today, so we get some numbers to discuss.

Namek commented:
Just a reference: #283

snorrees commented:
## Changes
* Some bug-fixes so CachedInjector works as intended.
* com.badlogic.gdx.utils.ObjectMap for caches instead of java.lang.Map (this is huge for performance)
* world.tryInject(object) - nonthrowing alternative to world.inject which is something I added for my own benifit (cache backed check for @Wire annotation)

## artemis-performance-test
I dont know if the new artemis-performance-test module should really be part of this, but build artemis, then buil the new module and run the run-benchmark.sh script which should produce something like this:
```
$ ./run_benchmark.sh
# VM invoker: C:\Program Files\Java\jre1.8.0_51\bin\java.exe
# VM options: <none>
# Warmup: 3 iterations, 2 s each
# Measurement: 10 iterations, 2 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: com.artemis.benchmark.InjectionBenchmark.createWorld_with_injectionCache

# Run progress: 0.00% complete, ETA 00:00:52
# Fork: 1 of 1
# Warmup Iteration   1: 108113.056 ops/s
# Warmup Iteration   2: 178407.104 ops/s
# Warmup Iteration   3: 181834.356 ops/s
Iteration   1: 184024.131 ops/s
Iteration   2: 182646.615 ops/s
Iteration   3: 179829.446 ops/s
Iteration   4: 183882.000 ops/s
Iteration   5: 179031.996 ops/s
Iteration   6: 184297.732 ops/s
Iteration   7: 186606.773 ops/s
Iteration   8: 184860.949 ops/s
Iteration   9: 180666.423 ops/s
Iteration  10: 183748.579 ops/s


Result: 182959.465 ▒(99.9%) 3630.371 ops/s [Average]
  Statistics: (min, avg, max) = (179031.996, 182959.465, 186606.773), stdev = 2401.266
  Confidence interval (99.9%): [179329.093, 186589.836]


# VM invoker: C:\Program Files\Java\jre1.8.0_51\bin\java.exe
# VM options: <none>
# Warmup: 3 iterations, 2 s each
# Measurement: 10 iterations, 2 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: com.artemis.benchmark.InjectionBenchmark.createWorld_without_injectionCache

# Run progress: 50.00% complete, ETA 00:00:28
# Fork: 1 of 1
# Warmup Iteration   1: 10438.822 ops/s
# Warmup Iteration   2: 23408.134 ops/s
# Warmup Iteration   3: 24075.272 ops/s
Iteration   1: 25478.607 ops/s
Iteration   2: 26250.757 ops/s
Iteration   3: 26284.012 ops/s
Iteration   4: 25651.641 ops/s
Iteration   5: 22788.488 ops/s
Iteration   6: 24249.317 ops/s
Iteration   7: 26498.025 ops/s
Iteration   8: 27520.010 ops/s
Iteration   9: 27611.355 ops/s
Iteration  10: 26368.550 ops/s


Result: 25870.076 ▒(99.9%) 2198.839 ops/s [Average]
  Statistics: (min, avg, max) = (22788.488, 25870.076, 27611.355), stdev = 1454.396
  Confidence interval (99.9%): [23671.237, 28068.914]


# Run complete. Total time: 00:00:57

Benchmark                                                       Mode  Samples       Score      Error  Units
c.a.b.InjectionBenchmark.createWorld_with_injectionCache       thrpt       10  182959.465 ▒ 3630.371  ops/s
c.a.b.InjectionBenchmark.createWorld_without_injectionCache    thrpt       10   25870.076 ▒ 2198.839  ops/s

```
Keep in mind that this is heavly in favor of the cache, as there is 3 warm-up rounds.

Tested manuall on android - from chat:

android load "Initialize entity templates" loadtask for the game "Metal Core" on Note 4 - timed from the application, reported in the log
This is real world numbers - there is no warmup, as this is the actually usecase.

    with CachedInjector v1.1 (fixes + ObjectMap): 1.2 seconds
    with OldschoolInjector: 10.40 seconds (!)

The loadtask calls

    injector.supportsInjection: 9200 times
    injector.inject : 7629 times

The ObjectMap sizes after all this:

    classCache: 105
    fieldClassTypeCache: 88
    namedWireCache: 536
    mapperTypeCache: 186

junkdog commented:
somewhat OT: you might want to increase the warmup duration by another 2s or so; the first real benchmark improves on the warmup throughput, ideally everything should be JITted by the time the actual benchmarks start.

Really liking those numbers btw. Does that mean a new beta is coming soon?

DaanVanYperen commented:
I wish libGDX would package common utility classes separately.

junkdog commented:
yeah, or see if there are even better alternatives, but borrowing libgdx's is prob the way forward.

snorrees commented:
I noticed that I had put a world.process() at the end of the benchmark - which I think is just noize for this. Removed it and increased warmup to 3 x 10 seconds. 

I added a hashmap implmentation just for comparison, the numbers for hashmap are **not as abysmal** as I first thought after fixing all the worst bug (Field equals & hashcode):
```
Benchmark                                                            Mode  Samples       Score       Error  Units
c.a.b.InjectionBenchmark.createWorld_with_hashmap_injectionCache    thrpt       10  201965.076 ▒ 14087.981  ops/s
c.a.b.InjectionBenchmark.createWorld_with_injectionCache            thrpt       10  210567.571 ▒  6893.526  ops/s
c.a.b.InjectionBenchmark.createWorld_without_injectionCache         thrpt       10   27402.705 ▒  2376.483  ops/s
```

Definitely a new build coming this weekend :grin: Singleton-world has proven very good for loadtimes!

DaanVanYperen commented:
@snorrees looking very good! I'd certainly give you a cookie if you manage to get the same results in core reflection. 

It might make sense if we maintain a drop in libgdx-stripped project. Gets rid of the double reflection cache for one. Then again got to keep in sync..

snorrees commented:
To make this useful for a merge should I:
* **A**: Change PR to only include changes necessary to make Injector configurable. 
    * I would preferably like to keep the tryInjection method on world, or something similar.
    * Make a PR to artemis-contrib with the CachedInjector
* **B**: Fix the commented stuff, and leave it as is.
* **C**: Fix the commented stuff, but remove the performance test module. 
* **D**: Scrap it - no point without expanding ClassReflection

junkdog commented on artemis/src/main/java/com/artemis/Inject.java:
```diff
@@ -0,0 +1,16 @@
+
+    void inject(Object target) throws RuntimeException;
+
+    boolean injectionSupported(Object target);
```
s/injectionSupported/supportsInjection/ ?

snorrees replied:
Motivation: Cached @Wire access - used by World before doing injection.

junkdog replied:
Ah, I just meant the name - `boolean supportsInjection(Object)` sounds better, no? (If no, then keep the name)

junkdog commented:
Just as I was running out of time, I discovered [Invisible Inc](http://store.steampowered.com/app/243970/). Bad timing...

So, I've only looked at the diff. I think we might need a GWT test, knowing how prone it is to generally just falling apart. But this probably means that we need to start using the emulated GWT relection wrappers.

We could prob stick to using HashMap; I'm not really sure when it allocates memory, but it's probably good enough for the time being.

The benchmark module is better suited for https://github.com/junkdog/entity-system-benchmarks/ - it could just sit there, or we'll make it a part of the benchmark suite.

Do we need the old injector anymore?

junkdog commented:
Not sure I'm liking the name of the interface - `Inject` reminds me too much of `@Inject`.

snorrees commented:
The interface will be named Injector. I only called Inject to make the diff easier.

junkdog commented:
> The interface will be named Injector

:+1:

snorrees commented:
And before throwing out the old Injector: I should probably make a benchmark which only tests cold-cache. The current benchmark has 100% cache-hit since it does warmup. For projects which only does injection once, on mostly single-instance classes, the CachedInjector will probably be slightly slower.

junkdog commented:
True, but only marginally and it's a one-time hit (I attribute my optimism to sleep deprivation; I'm sure I'll recover, [as others have done](https://github.com/junkdog/entity-system-benchmarks/)).

DaanVanYperen commented:
Since you've done so much work already, could I bribe you to include #283 as well as a minor addition? 

Basically looking for a custom resolver for fields that the injector can't resolve by itself. something like

```java
interface FieldResolver { 
   public Object resolve(Class targetClass, TargetFieldAndWhatElse criteria);
}
```

```java
injector.setResolver(myResolver)
```

My usecase: injecting custom loggers, extended component mappers etc.

snorrees commented:
> FieldResolver 
@DaanVanYperen Should this run even when the field is not annotated with @Wire?

DaanVanYperen commented:
> @DaanVanYperen Should this run even when the field is not annotated with @Wire?

Poor @Wire, he must hate his nickname. :p 

Good point, that might be a pain. hrm.

DaanVanYperen commented:
I'm not a big fan of introducing a new paradigm there, but having to @Wire is a pain. Is there a performance cost if we make @Wire default? (add a @SkipWire instead.).

snorrees commented:
How about this:
A default resolver which handles @Wire fields, and something like this:

```
interface FieldResolver { 
   Object resolve(Class targetClass, Field unresolvedField, FieldResolver defaultResolver);
}
```

@Wire - you just have to deal. Sorry.

DaanVanYperen commented:
`@Wire` on class implies a full wiring, actually wiring non native fields would be more intuitive in this case.

DaanVanYperen commented:
> How about this

Looks good. Did Field contain generics info? For MyMagicalComponentMapper<Pos> cases.

snorrees commented:
I dont know if I like full class wire, as often there is some other state involved.

You can get generic-info, for instance via ClassReflection, but its very slow. The question is if the FiledResolver should get access to some cachedClassDefinition or if thats out of scope.

DaanVanYperen commented:
How do we inject ComponentMapper now? It's basically what I want to do but for my own ComponentMapper derivative.

full class wire would be limited to fields that the injector actually knows how to inject.. It would just skip unresolvable fields.

snorrees commented:
Its a type-check for Mappers, Factories, Systems and Managers, falling back to a `@Wire` check, then throwing an exception IIRC.

The whole `@Wire` check can be skipped in favour of resolver.resolve(), so in that case you could create your own semantics for unresolved fields. 

In this case we dont have to change the annotation, use a defaultResolver which imitates todays behaviour, leaving the door open for you to do whatever you want.

Question is if costumResolver should use defaultResolverDelegate og just subclass the default implemetnation.

DaanVanYperen commented:
> The whole @Wire check can be skipped in favour of resolver.resolve(), so in that case you could create your own semantics for unresolved fields.

Good idea.That makes sense. You could model it as a chain like you suggested, nice and neat. So vanilla resolver -> custom resolver -> @wire resolver (if the user decides to chain).

snorrees commented:
Injector now has a FieldHandler, which manages a list of FieldResolvers. These resolvers are run in sequence until one returns a non-null value, which is then used by the injector to set the field value.

Artemis provides ArtemiesCoreFieldResolver & WiredFieldResolver, maintaing the same behaviour as before.

See FieldHandlerTest for usage.

See
